### DISCUSSION TITLE
Repeat Press Keys on Hold

### DISCUSSION THREAD

**Discussion by Danie10** [streamdeck-linux-gui/streamdeck-linux-gui#17] [Ideas / Feature Requests]
This is an essential feature for many games, and probably the one thing I really miss so far on the Linux driver for Stream Deck. So games like flight sims, Snowrunner, etc where you operate cranes, flaps, rudders, steering, etc. But may also be needed for any smooth sliding of adjustments for lighting levels, filters, etc versus right now repeated single presses of the deck's button.

Basically, a button needs to detect being held, and then repeat that programmed key press until released. I think it is mainly applicable to the "Press Keys" function, and unlikely to be for other button functionality like Commands. It may be as simple as adding some wrapper for IF THEN ELSE around the function to Press Keys.

This was issue #199 on previous project at https://github.com/timothycrosley/streamdeck-ui/issues/199 which contains more comments by others.

**Comment by Data-007** [comment]
Hi
I hope this may help, one of our programmers maybe able to look at the code and decipher it.
Look at the graphic in the readme section it seems to say by looking at it that this one can detect pressed keys.
https://github.com/abcminiuser/python-elgato-streamdeck
HIH
Data

**Reply by coolapso** [reply]
That is already in use on the code base. 

The whole project sits on top of that library, this application wouldn't exist.

**Comment by Danie10** [comment]
@Data-007 just want to check we are on the same feature? The image you linked to looks more like the button display reacts to a hold? I was thinking of if we hold a button, it keeps repeating that keypress signal. Right now a single press, or a long press, only seems to trigger one key combination.

So a bit like if I press and hold f on the keyboard, it sends ffffffffffffffffffffffffffffff through as long as I'm holding it. I'm thinking gamers like me would use this to control the swing of a crane, lifting/lowering a winch, etc while holding the key in.

**Reply by Data-007** [reply]
Hi @Danie10 
I was thinking if it can detect that the button is pressed, and trigger an change in graphic on the button, then
surly its not a stones throw away from adding in a repeating character(fffff) come signal of some description.

Not sure who to tag to ask but we have some talented individuals, three of which I know of. All of which are
on the discord.  Any off you guys like to take this on? Danie10 are you on discord? Could use the com's their
to ask. 

--

Data

**Comment by Danie10** [comment]
Ho @Data-007 yes I'm on the Discord channel, but I understood they asked we log on Github so it can be properly managed and assigned. Thing is I'm not sure actually that any of the new ideas in discussions here have been assigned or actioned yet. Maybe users are supposed to vote up the ideas they like so devs can prioritise?

The Windows Streamdeck app does do the repeat keypress sending, so technically it is quite possible.

**Reply by coolapso** [reply]
Just wrote something about this kind of topics, hope it clarifies your doubts: https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/discussions/41

**Comment by Data-007** [comment]
Hi all.
That post has cleared things up a lot. I not as proficient as some of you are aware of, not in some/all
things laid out in #41  I would lend assistance to anyone who is willing to take(voluntarily) some thing like this on.
I am sorry if I misled anyone it seems your right @Danie10  here it seams is the correct place.

I know it fess-able to do it as I said earlier in this thread, however fathoming out what to change and how is beyond me.
For now could a simple script of functionality like fix this, or would that to me hampered by only being triggered, once per press.
You can call scripts from the command box. 

HIH

**Reply by Danie10** [reply]
@Data-007 no problem. Yes, I'm just trying to clarify the requirement so that hopefully it catches the attention of one of the devs who may want to see a similar feature. It may also be that many of our existing potential devs are video streamers, and there may be gamer devs who for example may be able to come in here, and address this one specific feature request as it is likely more of value to gamers.

I'm a very rusty ex-dev myself, but I do suspect that implementing the basic repeat keypress on hold may just need an IF THEN ELSE wrapper around some existing code.

**Comment by Data-007** [comment]
> I'm a very rusty ex-dev myself, but I do suspect that implementing the basic repeat keypress on hold may just need an IF THEN ELSE wrapper around some existing code

Right I see, where I would stumble is working out where the key-press code is in the project.
An if then else wrapper, a simple script python should be easy to handle the if/then/else in a script hmm (thinking)

**Comment by bcdonadio** [comment]
Before repeating keys, I believe we should try to simply detect the KEY_DOWN and KEY_UP events that [are already sent by the core library](https://github.com/abcminiuser/python-elgato-streamdeck/pull/90). This enables both the PTT use case and enables one to write a much simpler hook to implement this key repeater logic without too many acrobatics.

**Reply by Data-007** [reply]
Hi @bcdonadio I like what I see/hear, here. I am some what puzzled as to how to do this, if you have any idea how to 
detect events like what you suggest then go for it I know what you mean but how to detect it I don't readily know.

I proposed a simple plugin feature see discussions in the project, one of the jobs would be to detect such behavior. #42

**Reply by coolapso** [reply]
totally agree @bcdonadio ...

**Reply by RoarkGit** [reply]
Agreed, having the ability to hold for PTT would be super nice and is really the only thing I'm missing personally..

**Comment by coolapso** [comment]
The goal of this feature request is clear,  there's too much talk about nonsense ... keep the topic related with technical stuff!

first its never just a if / else! there's always a lot more around the feature itself, error handling, maybe adding some rate rate limiting on how fast the keys are repeated (maybe a setting to adjust  that), There's tests that need to be added to the unit tests to make sure the functionalities work and do not break later on by some change. 

second there's a whole library that we depend on and whatever features need to be supported by the library as @bcdonadio  mentioned. 


if you don't really know how it works just leave it for those that actually know otherwise the thread gets to extensive and loses its purpose. 

It's okay to just upvote the feature request, react with and emoji, suggest things like what I had just said.
for example:

"I'd love to get this feature, would be nice to actually add a speed limiter to it as well so we can control how fast a key is repeated while pressing it"


Thank you all.

**Reply by coolapso** [reply]
This feature needs to be coordinated with: https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/discussions/26 as they might conflict!

**Comment by TheCyberQuake** [comment]
#177 Gives a POC of how we could pretty easily implement button holding.
Based on OP's replies and their original feature request on the original project, they don't want a turbo button they just want to be able to press and hold.
Easiest way for me to implement was just passing state boolean to keyboard_press_keys. From there keyboard_press_keys checks how many sections, and if there is only one section we can assume it's a single button and we can just use the passed state to call the new keyboard_key_down and keyboard_key_up to match the current button state.

**Reply by Danie10** [reply]
@TheCyberQuake thank you. Yes for gaming the log presses are "usually" a single key such as either A, W, S, Z, etc. Many users would hold CTRL, ALT, SHIFT keys when doing a key combo. So maybe starting out for single keys excluding CTRL, ALT and SHIFT keys would be a good start.

**Reply by TheCyberQuake** [reply]
At the moment I believe my POC should also support most key combos as well, though it's not fully tested.

**Reply by bcdonadio** [reply]
A specialized need of an user (weird key combo, game controller emulation, rotary media controller, etc...) can be provided if a hook with support for calling different external binaries is implemented at the arrival of a key up or a key down event. Basically, offloading the burden of implementing the mean to trigger whatever the user wishes to him/herself can be the first implementation in case the simple, already written evdev approach is not enough.

Later on, with greater project maturity, support for less common use cases can be implemented in a point-and-click way.

**Reply by bcdonadio** [reply]
Undefined or simply broken behavior is quite possible, for example, if you keep pressing/emulating a key combo while trying to type.

Keeping Ctrl+Alt+(some letter key) pressed to do whatever in some game may trigger a VT change if you press a F-key, a reboot if you press Delete, or the immediate bloodshed of all processes pertaining to the current user of the VT if you press Backspace.

Keeping Alt pressed will mess with your in-game chat, and can trigger quite low level subsystem resets or shutdowns if you throw PrintScreen into the mix because that's how you call [System Requests](https://en.wikipedia.org/wiki/Magic_SysRq_key) as well.

Interfacing with humans is troublesome. It is even harder if you need a computer to do it. :)

**Reply by TheCyberQuake** [reply]
Not really an issue since how I currently implemented is not a toggle, it will just hold the key while the user is holding the button. To be concerned with what other keys could be pressed while the user is holding the stream deck button would be of concern if they were holding that key combo on the keyboard as well.
I'm not entirely sure if I understand the point you are trying to make since to me it doesn't really seem relevant or of actual concern

**Reply by bcdonadio** [reply]
Don't worry, I most likely went into a tangent of a non-issue there, apologies.

I was trying to incentivize implementing it in the simplest way possible, but without necessarily tying that to a remap of another key press and go for "execute an arbitrary command" instead.

Cheers.
